test/test_sparse_csr.py

```diff
@@ -171,46 +171,131 @@ def genTensor(self, size, nnz, *, layout, device=None, dtype=torch.float, index_
             device = self.device_type
         return self.genSparseCompressedTensor(size, nnz, device=device, dtype=dtype, index_dtype=index_dtype, layout=layout)
 
-    def _generate_small_inputs(self, layout, device, dtype, index_dtype):
+    def _generate_small_inputs(self, layout, device, dtype, index_dtype,
+                               enable_batched=True,
+                               enable_hybrid=False):
         """Generator of inputs to sparse compressed tensor factory functions.
 
         The input is defined as a 4-tuple:
           compressed_indices, plain_indices, values, expected_size_from_shape_inference
         """
-        from operator import mul
-        from functools import reduce
-        if layout in {torch.sparse_csr, torch.sparse_csc}:
+
+        # TODO: set enable_hybrid default to True when Sparse
+        # Compressed tensors support dense dimensions
+
+        def shape(shape, basedim, blocksize=(1, 1)):
+            # Below, we define compressed and plain indices that
+            # correspond to row compressed tensors. In order to reuse
+            # the indices tensors for column compressed tensors, we
+            # swap the row and columns in shape dims (basedim and
+            # basedim + 1, respectively) to obtain the correct shape
+            # for column compressed tensors. Batch and dense
+            # dimensions remain as they are.
+            #
+            # Similarly, we reuse indices of non-block tensors for
+            # block tensors, that means, we'll need to multiply the
+            # base shape of the non-block tensor with blocksize to get
+            # the base shape of a block tensor.
+            if layout is torch.sparse_csc:
+                shape = shape[:basedim] + (shape[basedim + 1], shape[basedim]) + shape[basedim + 2:]
+            elif layout is torch.sparse_bsc:
+                shape = shape[:basedim] + (shape[basedim + 1] * blocksize[1], shape[basedim] * blocksize[0]) + shape[basedim + 2:]
+            elif layout is torch.sparse_bsr:
+                shape = shape[:basedim] + (shape[basedim] * blocksize[0], shape[basedim + 1] * blocksize[1]) + shape[basedim + 2:]
+            return shape
+
+        def values(lst, basedim, blocksize):
+            # Below, we define values for non-blocked tensors. To
+            # reuse these for blocked tensors, we replace all values
+            # in lst with a double-list that "shape" corresponds to
+            # blocksize.
+            if layout in {torch.sparse_bsr, torch.sparse_bsc}:
+
+                if layout is torch.sparse_bsc:
+                    blocksize = tuple(reversed(blocksize))
+
+                if not lst:
+                    return torch.tensor(lst, device=device, dtype=dtype).reshape(0, *blocksize)
+
+                def list_add(lst, value):
+                    if isinstance(lst, list):
+                        return [list_add(item, value) for item in lst]
+                    return lst + value
```

Comment:
Can you have list of lists?

Reply:
yes, this is exercised in BSR/BSC-related tests.